### PR TITLE
BackupFileDisclosure Handle Empty "Backup" Responses

### PR DESCRIPTION
**Pull request by kingthorin**
BackupFileDisclosure > Ignore empty "backup" responses.
ZapAddOn.xml > Add to changes section.
BackupFileDisclosureUnitTest > Added a test to cover the new behavior.

Related to zaproxy/zaproxy#4850

### CODE REVIEW

File: src/org/zaproxy/zap/extension/ascanrulesBeta/BackupFileDisclosure.java [hunk 1]
```diff
@@ -496,14 +496,16 @@ private void findBackupFile(HttpMessage originalMessage) throws Exception {
 				sendAndReceive(requestmsg, false);
 				disclosedData = requestmsg.getResponseBody().getBytes();
 				int requestStatusCode = requestmsg.getResponseHeader().getStatusCode();
+				// If the response is empty it's probably not really a backup
+				boolean isEmptyResponse = requestmsg.getResponseHeader().getContentLength() == 0;
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesBeta/BackupFileDisclosure.java]: Could check the actual length of the body (e.g. `disclosedData.length`).

File: src/org/zaproxy/zap/extension/ascanrulesBeta/BackupFileDisclosure.java [hunk 2]
```diff
@@ -496,14 +496,16 @@ private void findBackupFile(HttpMessage originalMessage) throws Exception {
...
 
 				//just to complicate things.. I have a test case which for the random file, does NOT give a 404 (so gives404s == false) 
 				//but for a "Copy of" file, actually gives a 404 (for some unknown reason). We need to handle this case.
-				if ( 	( gives404s && requestStatusCode != HttpStatus.SC_NOT_FOUND) || 
+				if ( 	( gives404s && requestStatusCode != HttpStatus.SC_NOT_FOUND && !isEmptyResponse) ||
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesBeta/BackupFileDisclosure.java]: The empty check could be done before both conditions. Worth extracting a method to make it easier to read, can be used here and in the parent check.

**Comment by kingthorin** [comment]
Tweaked

**Comment by kingthorin** [comment]
Re the unit test I dunno how to have nano serve an empty response for a backup request...

**Comment by thc202** [comment]
Yes, that should do it.

File: src/org/zaproxy/zap/extension/ascanrulesBeta/BackupFileDisclosure.java [hunk 3]
```diff
@@ -499,11 +503,11 @@ private void findBackupFile(HttpMessage originalMessage) throws Exception {
 
 				//just to complicate things.. I have a test case which for the random file, does NOT give a 404 (so gives404s == false) 
 				//but for a "Copy of" file, actually gives a 404 (for some unknown reason). We need to handle this case.
-				if ( 	( gives404s && requestStatusCode != HttpStatus.SC_NOT_FOUND) || 
-						(	(!gives404s) &&
-								nonexistfilemsg .getResponseHeader().getStatusCode() != requestStatusCode &&
-							(! Arrays.equals(disclosedData, nonexistfilemsgdata)) 
-						)
+				if ( 	( !isEmptyResponse(disclosedData) && gives404s &&
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesBeta/BackupFileDisclosure.java]: Sorry, I was suggesting doing the empty check (once) before both expressions not before each.

**Comment by kingthorin** [comment]
Look good?

File: src/org/zaproxy/zap/extension/ascanrulesBeta/BackupFileDisclosure.java [hunk 4]
```diff
@@ -499,11 +503,11 @@ private void findBackupFile(HttpMessage originalMessage) throws Exception {
 
 				//just to complicate things.. I have a test case which for the random file, does NOT give a 404 (so gives404s == false) 
 				//but for a "Copy of" file, actually gives a 404 (for some unknown reason). We need to handle this case.
-				if ( 	( gives404s && requestStatusCode != HttpStatus.SC_NOT_FOUND) || 
-						(	(!gives404s) &&
-								nonexistfilemsg .getResponseHeader().getStatusCode() != requestStatusCode &&
-							(! Arrays.equals(disclosedData, nonexistfilemsgdata)) 
-						)
+				if ( !isEmptyResponse(disclosedData) && ( gives404s &&
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesBeta/BackupFileDisclosure.java]: Needs parentheses surrounding the right operand (to be `gives404s` and `!gives404s`).

Comment (thc202) [reply src/org/zaproxy/zap/extension/ascanrulesBeta/BackupFileDisclosure.java]: I mean `isEmptyResponse(disclosedData) && ( (gives404s...) || (!gives404s...) )` (some redundant parentheses removed for clarity).

**Comment by kingthorin** [comment]
Should be good now (fingers crossed :grin: ).

**Comment by thc202** [comment]
Thank you!

**Pull request merged**